File: c_examples/alias.c
/*
 * Copyright (c) 2018 PolySync.
 *
 * Aliasing that results in unsafe memory access.
 */

#include <stdlib.h>
#include <stdint.h>

/* Basic types disallowed. */
typedef signed int SINT_32;

/* Example data, arbitrary. */
typedef struct { int32_t tux; int32_t baz; int32_t bar; } example_s;

/* Corrupts input. */
static void black_box(example_s ** alias_to_alias)
{
    if (alias_to_alias != NULL)
    {
        /* Data corrupted. */
        *alias_to_alias += 2048;
    }
}

SINT_32 main(void)
{
    /* Example data. */
    example_s data = {.tux = 1, .baz = 2, .bar = 3};
    /* Mutable alias to example data. */
    example_s * alias = &data;
    /* Mutable alias to alias. */
    example_s ** alias_to_alias = &alias;
    /* All values must be used. */
    int32_t return_value = alias->tux + alias->baz + alias->bar;

    /* Corrupts alias to data. */
    black_box(alias_to_alias);

    /* All values must be used. */
    if ((return_value != 0) && (alias_to_alias != NULL))
    {
        /* Use alias after corruption. */
        alias->tux = 4;
        alias->baz = 5;
        alias->bar = 6;

        /* All values must be used. */
        return_value = alias->tux + alias->baz + alias->bar;
    }

    /* All values must be used. */
    return (SINT_32)return_value;
}

File: rust_examples/alias.rs
// An attempt to reproduce the dangerous behavior of the corresponding alias.c

struct Example {
    x: i32,
    y: i32,
    z: i32,
}

fn black_box( _: &Box<&&Example> ) {
    panic!("Memory corrupted.");
}

fn main() {
    // Assign an arbitrary value to a variable for aliasing.
    let mut a: Example = Example { x: 1, y: 2, z: 3 };
    let mut b: Box<&Example> = Box::new(&mut a);
    let c: Box<&&Example> = Box::new(&mut b);

    black_box(&c);

    b.x = 4;
    b.y = 5;
    b.z = 6;
}
